Add user permission table

AkshayPall opened the pull request:
Admin checks are also done on every time the property is used, instead
if it being stored statically as a field of the User.

akprasad commented on ambuda/admin.py:
```diff
@@ -21,7 +21,7 @@ def index(self):
 class BaseView(sqla.ModelView):
     def is_accessible(self):
-        return current_user.is_admin
+        return q.is_user_admin(current_user.id)
```
The API feels a little boilerplate-y. What if we keep the `is_admin` property and do the query in `auth.py` ?

akprasad commented on ambuda/models/user_permissions.py:
```diff
@@ -0,0 +1,33 @@
+from datetime import datetime
```
Since this is short, you can keep this in `models.auth`.

akprasad replied:
Ah, there's no model.auth -- but there should be! Let's rename `models.users` and combine it with this.

akprasad commented on ambuda/seed/lookup/user_permissions.py:
```diff
@@ -0,0 +1,37 @@
+from sqlalchemy.orm import Session
```
Generally, database seeding scripts write a small amount of fixed data to the database. This file is more analogous to a migration and should live in the migrations script, IMO.

akprasad commented on migrations/versions/0d3fce7c5341_create_user_permission_table.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def downgrade() -> None:
+    op.drop_table("user_permission")
```
user_permissions

akprasad commented on test/ambuda/test_models.py:
```diff
@@ -1,5 +1,11 @@
+from ambuda.queries import get_session
+
+
+def _cleanup(session, *objects):
```
I think there's need to cleanup because in our current testing approach, each new `client` test should recreate the database from scratch.

AkshayPall replied:
I don't think this currently happens. I noticed errors when trying to create a User with the same email etc. as the test above this. I could've made a different user however I wanted to clean up the side effects from the existing tests in case any other contributors run into this issue. We should investigate and address this issue though - each test should be unique.

akprasad replied:
Fair enough -- let's keep it if it helps the tests pass, and let's find a way around it in a follow-up PR.

akprasad commented on ambuda/models/user_permissions.py:
```diff
@@ -0,0 +1,33 @@
+from ambuda.models.base import Base, foreign_key, pk
+
+
+class UserPermissionValues(IntEnum):
```
Hmm, there are a few approaches to representing a small set of ~enumerated values in a db:

1. ENUM column -- generally inflexible if adding more values later
2. Lookup table -- create a new table and refer to it with foreign keys. This also lets us add metadata
3. ENUM in code, as is done here

I don't know what's best between (2) and (3), but we currenty do (2) for page statuses, and if there's not a strong reason for (3) I prefer consistency. Or in other words, something like 1+2+4 here:

https://zoharpeled.wordpress.com/2020/11/07/role-based-security-database-design/

akprasad replied:
(PS: to be clear, I don't know the right answer to this and am happy to go with this PR's implementation if there's some prior art for it.)

AkshayPall replied:
I agree with this choice for (1) consistency and for (2) easy expansion with additional roles. I pushed a new commit to reflect this.

akprasad commented on ambuda/queries.py:
```diff
@@ -169,7 +169,39 @@ def create_user(*, username: str, email: str, raw_password: str):
     return user
+
+
+def get_user_permission(user_id: int) -> Optional[db.UserPermission]:
```
This can be done with a SQLAlchemy relationship, see `pages` on `Project` for an example.

akprasad commented on ambuda/models/user_permissions.py:
```diff
@@ -0,0 +1,33 @@
+
+class UserPermission(Base):
+
+    """A specific page revision.
```
update comments

akprasad reviewed (changes requested):
Wonderful -- thank you!

akprasad commented:
I haven't forgotten this! I've just been mulling over this blog post:

https://techspot.zzzeek.org/2011/01/14/the-enum-recipe/

Normally I'd just say we should go for it, but we'll be adding a lot of these enumerations soon (project status, tagging status, text type, text language, ...) and I want to think this through before the future migration work explodes.

akprasad commented on ambuda/models/auth.py:
```diff
@@ -0,0 +1,90 @@
+    __tablename__ = "user_roles"
+
+    #: Primary key.
+    id = pk()  # TODO: make pk both columns
```
I tihnk you should be able to do this by adding `primary_key=True` to both IDs and deleting `id` ?

akprasad commented on ambuda/models/auth.py:
```diff
@@ -0,0 +1,90 @@
+    #: Description of the role
+    description = Column(String, nullable=False)
+    #: When this permission was added
+    creation_time = Column(DateTime, nullable=False, default=datetime.utcnow)
```
I've been using `created_at` since it's a little shorter. Let's standardize on that.

akprasad commented on ambuda/models/auth.py:
```diff
@@ -0,0 +1,90 @@
+    creation_time = Column(DateTime, nullable=False, default=datetime.utcnow)
+
+
+class User(UserMixin, Base):
```
I added a `created_at` column for User -- you might have to rebase.

akprasad commented on migrations/versions/0d3fce7c5341_create_user_permission_table.py:
```diff
@@ -0,0 +1,77 @@
+from datetime import datetime
+from sqlalchemy.orm import Session
+
+import ambuda.database as db
```
I recently learned that importing from prod code is not a best practice for Alembic. The rationale is that by depending on `db` here, our migration file depends on code that might change in the future, and it could even break entirely.

Instead, we should lock the specific model as of this point. You can grep for "Base" in `migrations` to see some examples of where I do this.

akprasad commented on test/ambuda/views/test_admin.py:
```diff
@@ -3,6 +3,11 @@ def test_admin_index__unauth(client):
     assert resp.status_code == 404
 
 
+def test_admin_index__auth(rama_client):
```
Thanks! :)

akprasad commented on ambuda/models/auth.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class SiteRole(Enum):
+    PROOFREADER = "Allow proofreading of texts prior."
```
I like the idea, but it feels odd to me to have a description be on the RHS of an ENUM, since the RHS is basically the ground truth that we use and test in our logic. Something like:

```
class SiteRole(str, Enum):
    #: Can mark pages as yellow
    P1 = 'p1'
    #: Can mark pages as green
    P2 = 'p2'
    #: Administrator rights
    ADMIN = 'admin'
```

feels more intuitive. I think descriptions are useful if the enum values here are more user-facing, but here we're mainly using them for permissions tracking -- I think just `id` and `name` are necessary for our needs, and we can keep the descriptions in comments.

akprasad reviewed (commented):
Thanks for waiting. After looking into it, the ENUM world is a big mess, and let's just stick with lookup tables to unblock.

I think these are my last round of comments. Once these are resolved, we should be good to go!

akprasad commented on ambuda/models/base.py:
```diff
@@ -20,6 +20,8 @@ def pk():
 
 
-def foreign_key(field: str):
+def foreign_key(field: str, primary_key: bool = False):
```
I think the `primary_key` option will be rare. Better to keep this function simple and inline the function body where you need it.

akprasad commented on ambuda/database.py:
```diff
@@ -7,4 +7,4 @@
 from ambuda.models.talk import *
 from ambuda.models.texts import *
-from ambuda.models.users import *
+from ambuda.models.auth import *
```
Nit: sort imports

akprasad commented:
Looks good!! Just rebase against `main` and I'll merge it in.

AkshayPall commented:
Follow ups:
- Check for proofreading permissions prior to
    - GETing proofreading view
    - POSTing proofread revisions

AkshayPall commented:
@akprasad Addressing alembic migration issue currently, will let you know when it's safe to merge

akprasad commented:
> Changes:
> 
>     * switched SiteRole to StrEnum and removed the description column from `Role`

`StrEnum` is an interesting idea. I like that it saves boilerplate. But if you'll indulge me, I think it's better for us to avoid it. It's a small point, but it touches on how I think about our code base, so I'll write a bit too much about it here to explain how I think about it.

`auto()` is not saving us much compared to using the literal values. For the sake of a bit of copy-paste convenience, we're bringing in a new library whose conventions aren't obvious to new contributors. For example, perhaps `auto()` here means a _lowercase_ string value -- the reader can't know until they read about `auto` in the context of `StrEnum` as opposed to the other clasess in that module.

`strenum` is also a new and complex dependency for relatively little new functionality. Part of working in open-source is that we have to depend on other libraries, and there's a risk-reward tradeoff based on how badly we need that functionality and how vetted and secure the repo is in the community. For example, I use `conllu` in `dcs_utils`, even though it's small/risky, because we needed the ability to parse `conllu` files and the gain was that we could rapidly add parse data to the library.

The last point is that I'm not sure coupling between the enum variable name and the stringified value will be safe going forward. I think as a basic principle we should be free to rename everything in the codebase without having to worry about breaking business logic. Now, this is an ideal that we can't always have -- for example, renaming a view function will cause Flask routing to break, and we have to update a lot of call sites manually. But I think we should follow it where we can. The LHS is for humans, and the RHS is for machines.

AkshayPall commented:
> > Changes:
> > ```
> > * switched SiteRole to StrEnum and removed the description column from `Role`
> > ```
> 
> `StrEnum` is an interesting idea. I like that it saves boilerplate. But if you'll indulge me, I think it's better for us to avoid it. It's a small point, but it touches on how I think about our code base, so I'll write a bit too much about it here to explain how I see it.
> 
> `auto()` is not saving us much compared to using the literal values. For the sake of a bit of copy-paste convenience, we're bringing in a new library whose conventions aren't obvious to new contributors. For example, perhaps `auto()` here means a _lowercase_ string value -- the reader can't know until they read about `auto` in the context of `StrEnum` as opposed to the other clasess in that module.
> 
> `strenum` is also a new and complex dependency for relatively little new functionality. Part of working in open source is that we have to depend on other libraries, and there's a risk-reward tradeoff based on how badly we need that functionality and how vetted and secure the repo is in the community. For example, I use `conllu` in `dcs_utils`, even though it's small/risky, because we needed the ability to parse `conllu` files and the gain was that we could rapidly add parse data to the library.
> 
> The last point is that I'm not sure coupling between the enum variable name and the stringified value will be safe going forward. I think as a basic principle we should be free to rename everything in the codebase without having to worry about breaking business logic. Now, this is an ideal that we can't always have -- for example, renaming a view function will cause Flask routing to break, and we have to update a lot of call sites manually. But I think we should follow it where we can. The LHS is for humans, and the RHS is for machines.
> 
> One alternative is something like this:
> 
> ```
> class MyEnum(str, Enum):
>   FOO = 'foo'
>   BAR = 'bar'
>   BIFF = 'biff'
> ```
> 
> It's _slightly_ more manual, but it's also completely transparent about what it's doing and relies on standard library functions that have good documentation and support.

I agree with minimizing dependencies. I had thought the minimal complexity would be okay but the security concerns make the limited benefit minimal. I've reverted the changes - should be good to merge now.

akprasad commented:
Once you incorporate these small changes, we should be good to go. The main change I made was to ensure that the migration works given other migrations I've made since we started the PR.

> git apply diff.txt

[diff.txt](https://github.com/ambuda-project/ambuda/files/9252809/diff.txt)
